src/ql/pmgr/pass_types/base.cc

```diff
@@ -1000,15 +1000,20 @@ void Base::handle_debugging(
     utils::Str in_or_out = after_pass ? "out" : "in";
     auto debug_opt = options["debug"].as_str();
     if (debug_opt == "yes") {
-        ir->dump_seq(
-            utils::OutFile(context.output_prefix + "_debug_" + in_or_out + ".ir").unwrap()
-        );
-        ir::cqasm::WriteOptions write_options;
-        write_options.include_statistics = true;
-        ir::cqasm::write(
-            ir, write_options,
-            utils::OutFile(context.output_prefix + "_debug_" + in_or_out + ".cq").unwrap()
-        );
+        if (!after_pass && ir->program.empty()) {  // don't dump if we haven't got a program yet
```

Comment:
Why this check on after_pass?
I can imagine that the compiler ends with a qasm_writer and then runs a pass that cleans the IR (because it after that wants to read something else into the IR) and in between the program is empty. So I propose to omit the after_pass check and keep the program is empty check.

Reply:
Originally, the dumping was unconditional, which didn't make sense before running the very first pass, since the IR tends to be empty then (at least if that pass is `io.cqasm.read`). 
The condition introduced here was meant to be the most restrictive and thus give the least interference. It may be noteworthy to someone that is debugging if the IR becomes empty along the way.

Reply:
Yes, but then it is sufficient that the test is just` ir->program.empty()`.
I.e. omit `!after_pass,` because `after_pass` is also a reason when the program is empty to enter the if.

Reply:
I still think that the more restrictive the condition for suppression is, the better. Your proposal would also suppress debug output after a pass runs, if the program is empty, while there might (generally speaking) be interesting data in `ir->platform` (not with the existing passes though)

Reply:
Just checked, and without the extra test an exception is raised, basically telling that the Program is empty:

```
E       RuntimeError: Unknown error: in pass reader, phase debugging.before: dereferencing empty Maybe/One object of type N2ql2ir7ProgramE
E       Stack trace (most recent call last):
E       #31   Object "python3.6", at 0x107d5a25f, in _PyFunction_FastCallDict + 575
E       #30   Object "python3.6", at 0x107d596c2, in _PyEval_EvalCodeWithName + 2866
E       #29   Object "python3.6", at 0x107d55c0a, in _PyEval_EvalFrameDefault + 25626
E       #28   Object "python3.6", at 0x107c7aa44, in PyObject_Call + 100
E       #27   Object "python3.6", at 0x107c7ad82, in _PyObject_Call_Prepend + 130
E       #26   Object "python3.6", at 0x107c7ac05, in _PyObject_FastCallDict + 181
E       #25   Object "python3.6", at 0x107d5a25f, in _PyFunction_FastCallDict + 575
E       #24   Object "python3.6", at 0x107d596c2, in _PyEval_EvalCodeWithName + 2866
E       #23   Object "python3.6", at 0x107d55a11, in _PyEval_EvalFrameDefault + 25121
E       #22   Object "python3.6", at 0x107d58ab8, in call_function + 536
E       #21   Object "python3.6", at 0x107d59f38, in fast_function + 552
E       #20   Object "python3.6", at 0x107d55a11, in _PyEval_EvalFrameDefault + 25121
E       #19   Object "python3.6", at 0x107d58ab8, in call_function + 536
E       #18   Object "python3.6", at 0x107d59f38, in fast_function + 552
E       #17   Object "python3.6", at 0x107d55a11, in _PyEval_EvalFrameDefault + 25121
E       #16   Object "python3.6", at 0x107d58ab8, in call_function + 536
E       #15   Object "python3.6", at 0x107d59f38, in fast_function + 552
E       #14   Object "python3.6", at 0x107d55a11, in _PyEval_EvalFrameDefault + 25121
E       #13   Object "python3.6", at 0x107d58ae7, in call_function + 583
E       #12   Object "python3.6", at 0x107cc83d5, in _PyCFunction_FastCallDict + 165
E       #11   Object "_openql.cpython-36m-darwin.so", at 0x10a65e6af, in _wrap_Compiler_compile_with_frontend(_object*, _object*) + 159
E       #10   Object "_openql.cpython-36m-darwin.so", at 0x10ab68541, in ql::api::Compiler::compile_with_frontend(ql::api::Platform const&) + 49
E       #9    Object "_openql.cpython-36m-darwin.so", at 0x10a9b4dec, in ql::pmgr::Manager::compile(ql::utils::tree::base::One<ql::ir::Root> const&) + 108
E       #8    Object "_openql.cpython-36m-darwin.so", at 0x10a9a0dfd, in ql::pmgr::pass_types::Base::compile(ql::utils::tree::base::One<ql::ir::Root> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) + 2749
E       #7    Object "_openql.cpython-36m-darwin.so", at 0x10a9a02f6, in ql::pmgr::pass_types::Base::run_sub_passes(ql::utils::tree::base::One<ql::ir::Root> const&, ql::pmgr::pass_types::Context const&) const + 150
E       #6    Object "_openql.cpython-36m-darwin.so", at 0x10a9a0932, in ql::pmgr::pass_types::Base::compile(ql::utils::tree::base::One<ql::ir::Root> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) + 1522
E       #5    Object "_openql.cpython-36m-darwin.so", at 0x10a99f4c8, in ql::pmgr::pass_types::Base::handle_debugging(ql::utils::tree::base::One<ql::ir::Root> const&, ql::pmgr::pass_types::Context const&, bool) + 1176
E       #4    Object "_openql.cpython-36m-darwin.so", at 0x10a87f334, in ql::ir::cqasm::write(ql::utils::tree::base::One<ql::ir::Root> const&, ql::ir::cqasm::WriteOptions const&, std::__1::basic_ostream<char, std::__1::char_traits<char> >&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) + 84
E       #3    Object "_openql.cpython-36m-darwin.so", at 0x10a87f3bb, in ql::ir::cqasm::write(ql::utils::tree::base::One<ql::ir::Root> const&, ql::utils::tree::base::One<ql::ir::Node> const&, ql::ir::cqasm::WriteOptions const&, std::__1::basic_ostream<char, std::__1::char_traits<char> >&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) + 43
E       #2    Object "_openql.cpython-36m-darwin.so", at 0x10a887bfc, in ql::ir::cqasm::Writer::visit_root(ql::ir::Root&) + 204
E       #1    Object "_openql.cpython-36m-darwin.so", at 0x10a8366ef, in ql::utils::tree::base::Maybe<ql::ir::Program>::deref() const + 255
E       #0    Object "_openql.cpython-36m-darwin.so", at 0x10a6b9f71, in ql::utils::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, ql::utils::ExceptionType) + 273

pybuild/openql/openql.py:2088: RuntimeError
```

Reply:
In the root node of the IR, program is a Maybe field.
That implies that every access to it must check whether it is empty, because that is an allowed condition.
src/ql/ir/cqasm/write.cc visit_root line 219 doesn't do that.
It just starts to visit the Program node.
So the bug is in visit_root, it should check emptiness of Program and if so, just skip visiting the program.

As a generalization, we could check that all maybe etc. nodes, when accessed, are checked for their emptiness in the code.

Reply:
you're right, I will change write.cc, and revert the change here (next week, when I'm back from the US)